modules/programs/firefox.nix

```diff
@@ -40,15 +40,44 @@ let
 
   profilesIni = generators.toINI { } profiles;
 
-  mkUserJs = prefs: extraPrefs: ''
-    // Generated by Home Manager.
-
-    ${concatStrings (mapAttrsToList (name: value: ''
-      user_pref("${name}", ${builtins.toJSON value});
-    '') prefs)}
+  mkUserJs = prefs: extraPrefs: bookmarks:
+    let
+      prefs' = lib.optionalAttrs ({ } != bookmarks) {
+        "browser.bookmarks.file" =
+          builtins.toString (firefoxBookmarksFile bookmarks);
+        "browser.places.importBookmarksHTML" = true;
+      } // prefs;
+    in ''
+      // Generated by Home Manager.
+
+      ${concatStrings (mapAttrsToList (name: value: ''
+        user_pref("${name}", ${builtins.toJSON value});
+      '') prefs')}
+
+      ${extraPrefs}
+    '';
 
-    ${extraPrefs}
-  '';
+  firefoxBookmarksFile = bookmarks:
+    let
+      mapper = _: entry: ''
+        <DT><A HREF="${entry.url}" ADD_DATE="0" LAST_MODIFIED="0"${
+          lib.optionalString (!builtins.isNull entry.keyword)
+          " SHORTCUTURL=\"${entry.keyword}\""
+        }>${entry.name}</A>
```

Comment:
I think this will need XML escaping?

Reply:
done according to this, did not find existing xml escape code. 

https://stackoverflow.com/questions/1091945/what-characters-do-i-need-to-escape-in-xml-documents

Reply:
I've created a Nixpkgs PR to add an `escapeXML` function: https://github.com/NixOS/nixpkgs/pull/140356